diff --git a/src/main/java/ru/practicum/shareit/booking/Booking.java b/src/main/java/ru/practicum/shareit/booking/Booking.java
index 2d9c666..4762915 100644
--- a/src/main/java/ru/practicum/shareit/booking/Booking.java
+++ b/src/main/java/ru/practicum/shareit/booking/Booking.java
@@ -1,7 +1,13 @@
 package ru.practicum.shareit.booking;
 
-/**
- * TODO Sprint add-bookings.
- */
+import lombok.Data;
+
+import java.time.LocalDate;
+
+@Data
 public class Booking {
+    private LocalDate start;
+    private LocalDate end;
+    private Long itemId;
+    private Long userId;
 }
diff --git a/src/main/java/ru/practicum/shareit/exception/ErrorHandler.java b/src/main/java/ru/practicum/shareit/exception/ErrorHandler.java
new file mode 100644
index 0000000..1997c50
--- /dev/null
+++ b/src/main/java/ru/practicum/shareit/exception/ErrorHandler.java
@@ -0,0 +1,23 @@
+package ru.practicum.shareit.exception;
+
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.ExceptionHandler;
+import org.springframework.web.bind.annotation.ResponseStatus;
+import org.springframework.web.bind.annotation.RestControllerAdvice;
+
+@RestControllerAdvice
+public class ErrorHandler {
+
+    @ExceptionHandler(NotFoundException.class)
+    @ResponseStatus(HttpStatus.NOT_FOUND)
+    public ErrorResponse handleNotFoundException(NotFoundException exception) {
+        return new ErrorResponse(exception.getMessage());
+    }
+
+    @ExceptionHandler(ValidationException.class)
+    @ResponseStatus(HttpStatus.INTERNAL_SERVER_ERROR)
+    public ErrorResponse handleValidationException(ValidationException exception) {
+        return new ErrorResponse(exception.getMessage());
+    }
+
+}
diff --git a/src/main/java/ru/practicum/shareit/exception/ErrorResponse.java b/src/main/java/ru/practicum/shareit/exception/ErrorResponse.java
new file mode 100644
index 0000000..7313afe
--- /dev/null
+++ b/src/main/java/ru/practicum/shareit/exception/ErrorResponse.java
@@ -0,0 +1,12 @@
+package ru.practicum.shareit.exception;
+
+import lombok.Getter;
+
+@Getter
+public class ErrorResponse {
+    private String error;
+
+    public ErrorResponse(String error) {
+        this.error = error;
+    }
+}
diff --git a/src/main/java/ru/practicum/shareit/exception/NotFoundException.java b/src/main/java/ru/practicum/shareit/exception/NotFoundException.java
new file mode 100644
index 0000000..508b545
--- /dev/null
+++ b/src/main/java/ru/practicum/shareit/exception/NotFoundException.java
@@ -0,0 +1,7 @@
+package ru.practicum.shareit.exception;
+
+public class NotFoundException extends RuntimeException {
+    public NotFoundException(String message) {
+        super(message);
+    }
+}
diff --git a/src/main/java/ru/practicum/shareit/exception/ValidationException.java b/src/main/java/ru/practicum/shareit/exception/ValidationException.java
new file mode 100644
index 0000000..59043da
--- /dev/null
+++ b/src/main/java/ru/practicum/shareit/exception/ValidationException.java
@@ -0,0 +1,7 @@
+package ru.practicum.shareit.exception;
+
+public class ValidationException extends RuntimeException {
+    public ValidationException(String message) {
+        super(message);
+    }
+}
diff --git a/src/main/java/ru/practicum/shareit/item/InMemoryItemStorage.java b/src/main/java/ru/practicum/shareit/item/InMemoryItemStorage.java
new file mode 100644
index 0000000..534a1d7
--- /dev/null
+++ b/src/main/java/ru/practicum/shareit/item/InMemoryItemStorage.java
@@ -0,0 +1,60 @@
+package ru.practicum.shareit.item;
+
+import org.springframework.stereotype.Component;
+import ru.practicum.shareit.exception.NotFoundException;
+import ru.practicum.shareit.item.model.Item;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+
+@Component
+public class InMemoryItemStorage implements ItemStorage {
+
+    private final Map<Long, Item> items = new HashMap<>();
+    private Long lastId = 0L;
+
+    @Override
+    public Item addItem(Item item) {
+        lastId++;
+        items.put(lastId, item);
+        item.setId(lastId);
+        return item;
+    }
+
+    @Override
+    public Item updateItem(Item item) {
+        items.put(item.getId(), item);
+        return item;
+    }
+
+    @Override
+    public Item getItemById(Long itemId) {
+        Item item = items.get(itemId);
+        if (item == null) {
+            String message = String.format("вещь с id=%d не найдена", itemId);
+            throw new NotFoundException(message);
+        }
+        return item;
+    }
+
+    @Override
+    public Collection<Item> getItemsByOwner(Long userId) {
+        return items.values().stream()
+                .filter(item -> item.getOwnerId().equals(userId))
+                .toList();
+    }
+
+    @Override
+    public Collection<Item> searchForItems(String text) {
+        String textUpperCase = text.toUpperCase();
+        return items.values().stream()
+                .filter(
+                        item -> item.getAvailable() &&
+                            (item.getName().toUpperCase().contains(textUpperCase) ||
+                                    item.getDescription().toUpperCase().contains(textUpperCase))
+                )
+                .toList();
+    }
+
+}
diff --git a/src/main/java/ru/practicum/shareit/item/ItemController.java b/src/main/java/ru/practicum/shareit/item/ItemController.java
index bb17668..90dcd7a 100644
--- a/src/main/java/ru/practicum/shareit/item/ItemController.java
+++ b/src/main/java/ru/practicum/shareit/item/ItemController.java
@@ -1,12 +1,65 @@
 package ru.practicum.shareit.item;
 
+import jakarta.validation.Valid;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PatchMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestHeader;
 import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestParam;
 import org.springframework.web.bind.annotation.RestController;
+import ru.practicum.shareit.item.dto.ItemCreateRequest;
+import ru.practicum.shareit.item.dto.ItemResponse;
+import ru.practicum.shareit.item.dto.ItemUpdateRequest;
+
+import java.util.Collection;
 
-/**
- * TODO Sprint add-controllers.
- */
 @RestController
 @RequestMapping("/items")
+@RequiredArgsConstructor
+@Slf4j
 public class ItemController {
+
+    private final ItemService itemService;
+
+    @PostMapping
+    public ItemResponse createItem(
+            @RequestHeader("X-Sharer-User-Id") Long userId,
+            @RequestBody @Valid ItemCreateRequest itemData
+    ) {
+        log.info("ItemController: получен запрос на создание вещи {} (userId={})", itemData, userId);
+        return itemService.createItem(itemData, userId);
+    }
+
+    @PatchMapping("/{itemId}")
+    public ItemResponse updateItem(
+            @RequestHeader("X-Sharer-User-Id") Long userId,
+            @PathVariable Long itemId,
+            @RequestBody @Valid ItemUpdateRequest itemData
+    ) {
+        log.info("ItemController: получен запрос на обновление данных вещи {} (itemId={}, userId={})", itemData, itemId, userId);
+        return itemService.updateItem(userId, itemId, itemData);
+    }
+
+    @GetMapping("/{itemId}")
+    public ItemResponse getItemById(@PathVariable Long itemId) {
+        log.info("ItemController: получен запрос на получение данных вещи (itemId={})", itemId);
+        return itemService.getItemById(itemId);
+    }
+
+    @GetMapping
+    public Collection<ItemResponse> getItemsByOwner(@RequestHeader("X-Sharer-User-Id") Long userId) {
+        log.info("ItemController: получен запрос на получение вещей владельца (userId={})", userId);
+        return itemService.getItemsByOwner(userId);
+    }
+
+    @GetMapping("/search")
+    public Collection<ItemResponse> searchForItems(@RequestParam String text) {
+        log.info("ItemController: получен запрос на поиск вещей по строке (text={})", text);
+        return itemService.searchForItems(text);
+    }
 }
diff --git a/src/main/java/ru/practicum/shareit/item/ItemMapper.java b/src/main/java/ru/practicum/shareit/item/ItemMapper.java
new file mode 100644
index 0000000..de4d233
--- /dev/null
+++ b/src/main/java/ru/practicum/shareit/item/ItemMapper.java
@@ -0,0 +1,45 @@
+package ru.practicum.shareit.item;
+
+import lombok.experimental.UtilityClass;
+import ru.practicum.shareit.item.dto.ItemCreateRequest;
+import ru.practicum.shareit.item.dto.ItemResponse;
+import ru.practicum.shareit.item.dto.ItemUpdateRequest;
+import ru.practicum.shareit.item.model.Item;
+
+@UtilityClass
+public class ItemMapper {
+
+    public ItemResponse mapToItemResponse(Item item) {
+        ItemResponse itemResponse = new ItemResponse();
+        itemResponse.setId(item.getId());
+        itemResponse.setName(item.getName());
+        itemResponse.setDescription(item.getDescription());
+        itemResponse.setAvailable(item.getAvailable());
+        return itemResponse;
+    }
+
+    public Item mapItemCreateRequestToItem(ItemCreateRequest itemCreateRequest) {
+        Item item = new Item();
+        item.setName(itemCreateRequest.getName());
+        item.setDescription(itemCreateRequest.getDescription());
+        item.setAvailable(itemCreateRequest.getAvailable());
+        return item;
+    }
+
+    public void updateItemFields(Item item, ItemUpdateRequest itemUpdateRequest) {
+
+        if (itemUpdateRequest.hasName()) {
+            item.setName(itemUpdateRequest.getName());
+        }
+
+        if (itemUpdateRequest.hasDescription()) {
+            item.setDescription(itemUpdateRequest.getDescription());
+        }
+
+        if (itemUpdateRequest.hasAvailavle()) {
+            item.setAvailable(itemUpdateRequest.getAvailable());
+        }
+
+    }
+
+}
diff --git a/src/main/java/ru/practicum/shareit/item/ItemService.java b/src/main/java/ru/practicum/shareit/item/ItemService.java
new file mode 100644
index 0000000..0d61a7c
--- /dev/null
+++ b/src/main/java/ru/practicum/shareit/item/ItemService.java
@@ -0,0 +1,21 @@
+package ru.practicum.shareit.item;
+
+import ru.practicum.shareit.item.dto.ItemCreateRequest;
+import ru.practicum.shareit.item.dto.ItemResponse;
+import ru.practicum.shareit.item.dto.ItemUpdateRequest;
+
+import java.util.Collection;
+
+public interface ItemService {
+
+    ItemResponse createItem(ItemCreateRequest itemData, Long ownerId);
+
+    ItemResponse updateItem(Long userId, Long itemId, ItemUpdateRequest itemData);
+
+    ItemResponse getItemById(Long itemId);
+
+    Collection<ItemResponse> getItemsByOwner(Long userId);
+
+    Collection<ItemResponse> searchForItems(String text);
+
+}
diff --git a/src/main/java/ru/practicum/shareit/item/ItemServiceImpl.java b/src/main/java/ru/practicum/shareit/item/ItemServiceImpl.java
new file mode 100644
index 0000000..ad25795
--- /dev/null
+++ b/src/main/java/ru/practicum/shareit/item/ItemServiceImpl.java
@@ -0,0 +1,74 @@
+package ru.practicum.shareit.item;
+
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.stereotype.Service;
+import ru.practicum.shareit.exception.NotFoundException;
+import ru.practicum.shareit.item.dto.ItemCreateRequest;
+import ru.practicum.shareit.item.dto.ItemResponse;
+import ru.practicum.shareit.item.dto.ItemUpdateRequest;
+import ru.practicum.shareit.item.model.Item;
+import ru.practicum.shareit.user.UserStorage;
+
+import java.util.ArrayList;
+import java.util.Collection;
+
+@Service
+@RequiredArgsConstructor
+@Slf4j
+public class ItemServiceImpl implements ItemService {
+
+    private final ItemStorage itemStorage;
+    private final UserStorage userStorage;
+
+    @Override
+    public ItemResponse createItem(ItemCreateRequest itemData, Long ownerId) {
+        log.info("ItemServiceImpl: начало добавления новой вещи {} пользователем {}", itemData, ownerId);
+        userStorage.checkIfUserExists(ownerId);
+        Item item = ItemMapper.mapItemCreateRequestToItem(itemData);
+        item.setOwnerId(ownerId);
+        log.info("ItemServiceImpl: добавлена новая вещь {} пользователем {}", itemData, ownerId);
+        return ItemMapper.mapToItemResponse(itemStorage.addItem(item));
+    }
+
+    @Override
+    public ItemResponse updateItem(Long ownerId, Long itemId, ItemUpdateRequest itemData) {
+        log.info("ItemServiceImpl: начало обновления данных вещи {} (ownerId={}, itemId={})", itemData, ownerId, itemId);
+        userStorage.checkIfUserExists(ownerId);
+        Item item = itemStorage.getItemById(itemId);
+        if (!item.getOwnerId().equals(ownerId)) {
+            String message = String.format("Пользователь %d не является владельцем вещи %d", ownerId, itemId);
+            throw new NotFoundException(message);
+        }
+        ItemMapper.updateItemFields(item, itemData);
+        itemStorage.updateItem(item);
+        log.info("ItemServiceImpl: обновлены данные вещи {} (ownerId={}, itemId={})", itemData, ownerId, itemId);
+        return ItemMapper.mapToItemResponse(item);
+    }
+
+    @Override
+    public ItemResponse getItemById(Long itemId) {
+        log.info("ItemServiceImpl: получение данных вещи по id (itemId={})", itemId);
+        return ItemMapper.mapToItemResponse(itemStorage.getItemById(itemId));
+    }
+
+    @Override
+    public Collection<ItemResponse> getItemsByOwner(Long userId) {
+        log.info("ItemServiceImpl: получение вещей пользователя (userId={})", userId);
+        return itemStorage.getItemsByOwner(userId)
+                .stream()
+                .map(ItemMapper::mapToItemResponse).toList();
+    }
+
+    @Override
+    public Collection<ItemResponse> searchForItems(String text) {
+        log.info("ItemServiceImpl: поиск вещей по строке (text={})", text);
+        if (text.isEmpty()) {
+            return new ArrayList<>();
+        }
+        return itemStorage.searchForItems(text).stream()
+                .map(ItemMapper::mapToItemResponse)
+                .toList();
+    }
+
+}
diff --git a/src/main/java/ru/practicum/shareit/item/ItemStorage.java b/src/main/java/ru/practicum/shareit/item/ItemStorage.java
new file mode 100644
index 0000000..d97dcb3
--- /dev/null
+++ b/src/main/java/ru/practicum/shareit/item/ItemStorage.java
@@ -0,0 +1,19 @@
+package ru.practicum.shareit.item;
+
+import ru.practicum.shareit.item.model.Item;
+
+import java.util.Collection;
+
+public interface ItemStorage {
+
+    Item addItem(Item item);
+
+    Item updateItem(Item item);
+
+    Item getItemById(Long itemId);
+
+    Collection<Item> getItemsByOwner(Long userId);
+
+    Collection<Item> searchForItems(String text);
+
+}
diff --git a/src/main/java/ru/practicum/shareit/item/dto/ItemCreateRequest.java b/src/main/java/ru/practicum/shareit/item/dto/ItemCreateRequest.java
new file mode 100644
index 0000000..ce796be
--- /dev/null
+++ b/src/main/java/ru/practicum/shareit/item/dto/ItemCreateRequest.java
@@ -0,0 +1,19 @@
+package ru.practicum.shareit.item.dto;
+
+import jakarta.validation.constraints.NotBlank;
+import jakarta.validation.constraints.NotNull;
+import lombok.Data;
+
+@Data
+public class ItemCreateRequest {
+
+    @NotBlank
+    private String name;
+
+    @NotBlank
+    private String description;
+
+    @NotNull
+    private Boolean available;
+
+}
diff --git a/src/main/java/ru/practicum/shareit/item/dto/ItemDto.java b/src/main/java/ru/practicum/shareit/item/dto/ItemDto.java
deleted file mode 100644
index 9319d7d..0000000
--- a/src/main/java/ru/practicum/shareit/item/dto/ItemDto.java
+++ /dev/null
@@ -1,7 +0,0 @@
-package ru.practicum.shareit.item.dto;
-
-/**
- * TODO Sprint add-controllers.
- */
-public class ItemDto {
-}
diff --git a/src/main/java/ru/practicum/shareit/item/dto/ItemResponse.java b/src/main/java/ru/practicum/shareit/item/dto/ItemResponse.java
new file mode 100644
index 0000000..6e3745c
--- /dev/null
+++ b/src/main/java/ru/practicum/shareit/item/dto/ItemResponse.java
@@ -0,0 +1,11 @@
+package ru.practicum.shareit.item.dto;
+
+import lombok.Data;
+
+@Data
+public class ItemResponse {
+    private Long id;
+    private String name;
+    private String description;
+    private Boolean available;
+}
diff --git a/src/main/java/ru/practicum/shareit/item/dto/ItemUpdateRequest.java b/src/main/java/ru/practicum/shareit/item/dto/ItemUpdateRequest.java
new file mode 100644
index 0000000..37d3353
--- /dev/null
+++ b/src/main/java/ru/practicum/shareit/item/dto/ItemUpdateRequest.java
@@ -0,0 +1,28 @@
+package ru.practicum.shareit.item.dto;
+
+import jakarta.validation.constraints.Pattern;
+import lombok.Data;
+
+@Data
+public class ItemUpdateRequest {
+
+    @Pattern(regexp = ".*\\S.*", message = "Название не должно быть пустым или состоять только из пробелов")
+    private String name;
+
+    @Pattern(regexp = ".*\\S.*", message = "Описание не должно быть пустым или состоять только из пробелов")
+    private String description;
+
+    private Boolean available;
+
+    public boolean hasName() {
+        return name != null;
+    }
+
+    public boolean hasDescription() {
+        return description != null;
+    }
+
+    public boolean hasAvailavle() {
+        return available != null;
+    }
+}
diff --git a/src/main/java/ru/practicum/shareit/item/model/Item.java b/src/main/java/ru/practicum/shareit/item/model/Item.java
index 44eb73d..e997d55 100644
--- a/src/main/java/ru/practicum/shareit/item/model/Item.java
+++ b/src/main/java/ru/practicum/shareit/item/model/Item.java
@@ -1,7 +1,12 @@
 package ru.practicum.shareit.item.model;
 
-/**
- * TODO Sprint add-controllers.
- */
+import lombok.Data;
+
+@Data
 public class Item {
+    private Long id;
+    private Long ownerId;
+    private String name;
+    private String description;
+    private Boolean available;
 }
diff --git a/src/main/java/ru/practicum/shareit/request/ItemRequest.java b/src/main/java/ru/practicum/shareit/request/ItemRequest.java
index 95d6f23..b557955 100644
--- a/src/main/java/ru/practicum/shareit/request/ItemRequest.java
+++ b/src/main/java/ru/practicum/shareit/request/ItemRequest.java
@@ -1,7 +1,10 @@
 package ru.practicum.shareit.request;
 
-/**
- * TODO Sprint add-item-requests.
- */
+import lombok.Data;
+
+@Data
 public class ItemRequest {
+    private Long id;
+    private String name;
+    private String description;
 }
diff --git a/src/main/java/ru/practicum/shareit/user/InMemoryUserStorage.java b/src/main/java/ru/practicum/shareit/user/InMemoryUserStorage.java
new file mode 100644
index 0000000..68be4ee
--- /dev/null
+++ b/src/main/java/ru/practicum/shareit/user/InMemoryUserStorage.java
@@ -0,0 +1,63 @@
+package ru.practicum.shareit.user;
+
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.stereotype.Component;
+import ru.practicum.shareit.exception.NotFoundException;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+@Component
+@Slf4j
+public class InMemoryUserStorage implements UserStorage {
+
+    private final Map<Long, User> users = new HashMap<>();
+    private Long lastId = 0L;
+
+    @Override
+    public User getUserById(Long id) {
+        log.info("InMemoryUserStorage: начало поиска пользователя по id={}", id);
+        User user = users.get(id);
+        if (user == null) {
+            String message = String.format("Пользователь с id=%d не найден", id);
+            throw new NotFoundException(message);
+        }
+        return user;
+    }
+
+    @Override
+    public User addUser(User user) {
+        log.info("InMemoryUserStorage: начало добавления пользователя {}", user);
+        lastId++;
+        user.setId(lastId);
+        users.put(lastId, user);
+        log.info("InMemoryUserStorage: успешно добавлен пользователь {}", user);
+        return user;
+    }
+
+    @Override
+    public User updateUser(User user) {
+        users.put(user.getId(), user);
+        return user;
+    }
+
+    @Override
+    public void deleteUser(Long id) {
+        log.info("InMemoryUserStorage: начало удаления пользователя по id={}", id);
+        users.remove(id);
+        log.info("InMemoryUserStorage: удален пользователь с id={}", id);
+    }
+
+    @Override
+    public Optional<User> getUserByEmail(String email) {
+        return users.values().stream()
+                .filter(user -> user.getEmail().equals(email))
+                .findFirst();
+    }
+
+    @Override
+    public void checkIfUserExists(Long id) {
+        getUserById(id);
+    }
+}
diff --git a/src/main/java/ru/practicum/shareit/user/User.java b/src/main/java/ru/practicum/shareit/user/User.java
index ae6e7f3..c2706ad 100644
--- a/src/main/java/ru/practicum/shareit/user/User.java
+++ b/src/main/java/ru/practicum/shareit/user/User.java
@@ -1,7 +1,10 @@
 package ru.practicum.shareit.user;
 
-/**
- * TODO Sprint add-controllers.
- */
+import lombok.Data;
+
+@Data
 public class User {
+    private Long id;
+    private String name;
+    private String email;
 }
diff --git a/src/main/java/ru/practicum/shareit/user/UserController.java b/src/main/java/ru/practicum/shareit/user/UserController.java
index 03039b9..466b82f 100644
--- a/src/main/java/ru/practicum/shareit/user/UserController.java
+++ b/src/main/java/ru/practicum/shareit/user/UserController.java
@@ -1,12 +1,52 @@
 package ru.practicum.shareit.user;
 
+import jakarta.validation.Valid;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.web.bind.annotation.DeleteMapping;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PatchMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
 import org.springframework.web.bind.annotation.RequestMapping;
 import org.springframework.web.bind.annotation.RestController;
+import ru.practicum.shareit.user.dto.CreateUserRequest;
+import ru.practicum.shareit.user.dto.UpdateUserRequest;
 
-/**
- * TODO Sprint add-controllers.
- */
 @RestController
 @RequestMapping(path = "/users")
+@RequiredArgsConstructor
+@Slf4j
 public class UserController {
+
+    private final UserService userService;
+
+    @GetMapping("/{id}")
+    public User getUser(@PathVariable Long id) {
+        log.info("UserController: получен запрос на получение пользователя по id={}", id);
+        return userService.getUserById(id);
+    }
+
+    @PostMapping
+    public User createUser(@RequestBody @Valid CreateUserRequest userData) {
+        log.info("UserController: получен запрос на добавление пользователя {}", userData);
+        return userService.createUser(userData);
+    }
+
+    @PatchMapping("/{id}")
+    public User updateUser(
+            @PathVariable Long id,
+            @RequestBody @Valid UpdateUserRequest userData
+    ) {
+        log.info("UserController: получен запрос на обновление данных пользователя (id={}) {}", id, userData);
+        return userService.updateUser(id, userData);
+    }
+
+    @DeleteMapping("/{id}")
+    public void deleteUser(@PathVariable Long id) {
+        log.info("UserController: получен запрос на удаление пользователя по id={}", id);
+        userService.deleteUser(id);
+    }
+
 }
diff --git a/src/main/java/ru/practicum/shareit/user/UserMapper.java b/src/main/java/ru/practicum/shareit/user/UserMapper.java
new file mode 100644
index 0000000..b66b3d1
--- /dev/null
+++ b/src/main/java/ru/practicum/shareit/user/UserMapper.java
@@ -0,0 +1,28 @@
+package ru.practicum.shareit.user;
+
+import lombok.experimental.UtilityClass;
+import ru.practicum.shareit.user.dto.CreateUserRequest;
+import ru.practicum.shareit.user.dto.UpdateUserRequest;
+
+@UtilityClass
+public class UserMapper {
+
+    public User mapCreateUserDtoToUser(CreateUserRequest userData) {
+        User user = new User();
+        user.setName(userData.getName());
+        user.setEmail(userData.getEmail());
+        return user;
+    }
+
+    public void updateUserFields(User user, UpdateUserRequest userData) {
+
+        if (userData.hasName()) {
+            user.setName(userData.getName());
+        }
+
+        if (userData.hasEmail()) {
+            user.setEmail(userData.getEmail());
+        }
+
+    }
+}
diff --git a/src/main/java/ru/practicum/shareit/user/UserService.java b/src/main/java/ru/practicum/shareit/user/UserService.java
new file mode 100644
index 0000000..6d5677d
--- /dev/null
+++ b/src/main/java/ru/practicum/shareit/user/UserService.java
@@ -0,0 +1,56 @@
+package ru.practicum.shareit.user;
+
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.stereotype.Service;
+import ru.practicum.shareit.exception.ValidationException;
+import ru.practicum.shareit.user.dto.CreateUserRequest;
+import ru.practicum.shareit.user.dto.UpdateUserRequest;
+
+import java.util.Optional;
+
+@Service
+@RequiredArgsConstructor
+@Slf4j
+public class UserService {
+    private final UserStorage userStorage;
+
+    public User getUserById(Long id) {
+        log.info("UserService: начало получения пользователя по id={}", id);
+        return userStorage.getUserById(id);
+    }
+
+    public User createUser(CreateUserRequest userData) {
+        log.info("UserService: начало создания пользователя {}", userData);
+        Optional<User> userFoundByEmailOpt = userStorage.getUserByEmail(userData.getEmail());
+        if (userFoundByEmailOpt.isPresent()) {
+            String message = String.format("Пользователь с email='%s' уже существует", userData.getEmail());
+            throw new ValidationException(message);
+        }
+        User user = UserMapper.mapCreateUserDtoToUser(userData);
+        log.info("UserService: создан пользователь {}", user);
+        return userStorage.addUser(user);
+    }
+
+    public User updateUser(Long id, UpdateUserRequest userData) {
+        log.info("UserService: начало обновления данных пользователя (id={}) {}", id, userData);
+        User user = userStorage.getUserById(id);
+        if (userData.hasEmail()) {
+            Optional<User> userFoundByEmailOpt = userStorage.getUserByEmail(userData.getEmail());
+            if (userFoundByEmailOpt.isPresent() && !userFoundByEmailOpt.get().getId().equals(id)) {
+                String message = String.format("Пользователь с email='%s' уже существует", userData.getEmail());
+                throw new ValidationException(message);
+            }
+        }
+        UserMapper.updateUserFields(user, userData);
+        userStorage.updateUser(user);
+        log.info("UserService: обновлены данные пользователя (id={}) {}", id, userData);
+        return user;
+    }
+
+    public void deleteUser(Long id) {
+        log.info("UserService: начало удаления пользователя по id={}", id);
+        userStorage.deleteUser(id);
+        log.info("UserService: удален пользователь с id={}", id);
+    }
+}
diff --git a/src/main/java/ru/practicum/shareit/user/UserStorage.java b/src/main/java/ru/practicum/shareit/user/UserStorage.java
new file mode 100644
index 0000000..297f76d
--- /dev/null
+++ b/src/main/java/ru/practicum/shareit/user/UserStorage.java
@@ -0,0 +1,19 @@
+package ru.practicum.shareit.user;
+
+import java.util.Optional;
+
+public interface UserStorage {
+
+    User getUserById(Long id);
+
+    User addUser(User user);
+
+    User updateUser(User user);
+
+    void deleteUser(Long id);
+
+    Optional<User> getUserByEmail(String email);
+
+    void checkIfUserExists(Long id);
+
+}
diff --git a/src/main/java/ru/practicum/shareit/user/dto/CreateUserRequest.java b/src/main/java/ru/practicum/shareit/user/dto/CreateUserRequest.java
new file mode 100644
index 0000000..466a79e
--- /dev/null
+++ b/src/main/java/ru/practicum/shareit/user/dto/CreateUserRequest.java
@@ -0,0 +1,16 @@
+package ru.practicum.shareit.user.dto;
+
+import jakarta.validation.constraints.Email;
+import jakarta.validation.constraints.NotBlank;
+import jakarta.validation.constraints.NotNull;
+import lombok.Data;
+
+@Data
+public class CreateUserRequest {
+    @NotBlank
+    private String name;
+
+    @NotNull
+    @Email
+    private String email;
+}
diff --git a/src/main/java/ru/practicum/shareit/user/dto/UpdateUserRequest.java b/src/main/java/ru/practicum/shareit/user/dto/UpdateUserRequest.java
new file mode 100644
index 0000000..f0ad75a
--- /dev/null
+++ b/src/main/java/ru/practicum/shareit/user/dto/UpdateUserRequest.java
@@ -0,0 +1,23 @@
+package ru.practicum.shareit.user.dto;
+
+import jakarta.validation.constraints.Email;
+import jakarta.validation.constraints.Pattern;
+import lombok.Data;
+
+@Data
+public class UpdateUserRequest {
+
+    @Pattern(regexp = ".*\\S.*", message = "Имя не должно быть пустым или состоять только из пробелов")
+    private String name;
+
+    @Email
+    private String email;
+
+    public boolean hasName() {
+        return name != null;
+    }
+
+    public boolean hasEmail() {
+        return email != null;
+    }
+}
